components/datetime/src/timezone.rs

```diff
@@ -0,0 +1,331 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::borrow::Cow;
+
+use crate::{
+    date::TimeZoneInput,
+    format::timezone::FormattedTimeZone,
+    pattern::Error as PatternError,
+    provider::{
+        self,
+        timezones::{
+            ExemplarCitiesV1, MetaZoneGenericNamesLongV1, MetaZoneGenericNamesShortV1,
+            MetaZoneSpecificNamesLongV1, MetaZoneSpecificNamesShortV1, TimeZoneFormatsV1,
+        },
+    },
+    DateTimeFormatError,
+};
+use crate::{format::timezone, mock::timezone::ZeroPadding};
+use icu_locid::{LanguageIdentifier, Locale};
+use icu_provider::{DataProvider, DataRequest, ResourceKey, ResourceOptions, ResourcePath};
+
+use crate::fields::{FieldSymbol, TimeZone};
+use crate::pattern::{Pattern, PatternItem};
+
+/// Produces a vector containing the time-zone resource keys required to format a pattern.
+fn required_time_zone_keys(pattern: &Pattern) -> Result<Vec<ResourceKey>, DateTimeFormatError> {
```

Comment:
Issue: The idea with data keys is that they can be sliced away via dead code elimination.  After you build your app, you can search for data keys that are present, and use that information to reduce your data bundle size.  Rust is smart, but I doubt that it is smart *enough* to eliminate ResourceKeys that are stored in vectors.

Suggestion: Reformulate this function into multiple smaller `#[inline]` functions that are question-and-answer style.  "Do I need specific short names, yes or no? Do I need exemplar cities, yes or no?"  This should be enough (I think) to help the compiler do the dead code stripping.

Reply:
Thanks for explaining that. 

I actually just went with a simple approach of iterating through the relevant symbols in-line inside the `try_new()` function and just loading the resources directly. 

It makes the function a bit long, but I think it's still plenty readable.